src/parse.rs

```diff
@@ -679,6 +681,78 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn parse_conditional(&mut self, ix: usize, depth: usize) -> Result<(usize, Expr)> {
+        let bytes = self.re.as_bytes();
+        let b = bytes[ix];
+        let (next, condition) = if is_digit(b) {
+            self.parse_numbered_backref(ix)?
+        } else if b == b'\'' {
+            self.parse_named_backref(ix, "'", "'")?
+        } else if b == b'<' {
+            self.parse_named_backref(ix, "<", ">")?
+        } else {
+            self.parse_re(ix, depth)?
+        };
+        let b = bytes[next];
+        if b != b')' {
+            return Err(Error::ParseError(
+                next,
+                ParseError::GeneralParseError(
+                    "next byte isn't a closing paren after condition".to_string(),
+                ),
+            ));
+        }
+        let (end, child) = self.parse_re(next + 1, depth)?;
+        let b = bytes[end];
+        if b != b')' {
+            return Err(Error::ParseError(
+                end,
+                ParseError::GeneralParseError(
+                    "next byte isn't a closing paren at after condition".to_string(),
+                ),
+            ));
+        }
+        if end == next + 1 {
+            // Backreference validity checker
+            if let Expr::Backref(group) = condition {
+                return Ok((end + 1, Expr::BackrefExistsCondition(group)));
+            } else {
+                return Err(Error::ParseError(
+                    end,
+                    ParseError::GeneralParseError(
+                        "expected conditional to be a backreference or at least an expression for when the condition is true".to_string()
+                    )
+                ));
+            }
+        }
+        let if_true: Expr;
+        let mut if_false: Expr = Expr::Empty;
+        if let Expr::Alt(mut alternatives) = child {
+            // the truth branch will be the first alternative
+            if_true = alternatives.remove(0);
+            // if there is only one alternative left, take it out the Expr::Alt
+            if alternatives.len() == 1 {
+                if_false = alternatives.pop().expect("expected 2 alternatives");
+            } else {
+                // otherwise the remaining branches become the false branch
+                if_false = Expr::Alt(alternatives);
+            }
+        } else {
+            // there is only one branch - the truth branch. i.e. "if" without "else"
+            if_true = child;
+        }
+        let inner_condition = if let Expr::Backref(group) = condition {
+            Expr::BackrefExistsCondition(group)
+        } else {
+            condition
+        };
+
+        Ok((
+            end + 1,
+            Expr::Alt(vec![Expr::Concat(vec![inner_condition, if_true]), if_false]),
```

Comment:
Unfortunately there is a bug with this approach - the idea of being able to use `Expr::Alt` for the inner condition only works if the inner condition is the `BackrefExistsCondition`. This is why the test `assert_no_match(r"^(?(\d)abc|\d!)$", "5!");` fails. `\d` from the condition matches, but `abc` doesn't, so it should fail. But due to the alternation, it tries `\d!`, which passes. Instead, the `\d` from the condition should have consumed the `5` already. I will try to think how best to solve this, maybe with Split somehow?

Reply:
Okay, so I worked out that the example in my previous comment wasn't the best, as only lookarounds or backref validity checks can be used in the conditional expression. I have updated the code and tests accordingly. But I still need to fix it to ensure we don't try the false branch when the truth branch fails after the condition succeeded.

Reply:
> Okay, so I worked out that the example in my previous comment wasn't the best, as only lookarounds or backref validity checks can be used in the conditional expression.

I was wrong again :sweat_smile: but it's not clear to me how it should work with the IX if the condition consumes some characters (i.e. is a non-lookaround) and fails halfway through. Does the false branch execute from that position, or does it need to get reset to what it was before the condition was evaluated? I couldn't work it out from the Oniguruma test cases.

Reply:
Hmm would this be enough for a test of that behavior?: `(?(ab)c|d)`. It matches only `d` in `ad`, not `ad`.

So that would suggest the second option, it is reset. That's also what I would have expected after reading the info page and before trying it out.

Reply:
Thanks for checking. :) So we could treat it as an implicit lookahead then? This would match what RegexBuddy was implying. I'll try that when I get time.

Reply:
Or we only need to reset when the false branch is hit? I guess I could also try Oniguruma locally with some additional test cases for exploration purposes. Especially as RegexBuddy doesn't seem to have an engine/mode specifically for Oniguruma.

Reply:
Silly question, how are you executing arbitrary regex patterns against Oniguruma?

None of [the samples on their repo](https://github.com/kkos/oniguruma#sample-programs) seem to allow passing in a pattern and text as arguments on the command line, surely I'm missing something simple...

Reply:
I normally use the rust onig crate's API. I think they also have examples that allow you to execute patterns: https://github.com/rust-onig/rust-onig

Reply:
(BTW FYI I'm going to be on holidays for some time, so might be slow to respond :))

Reply:
Thanks for the tip - that helped a lot :) I have pushed some more changes and test cases which helps align with Oniguruma further. Have a great holiday! I'm in no hurry for feedback, so please enjoy/relax and don't worry about this PR :)